### PR TITLE
422 issue

### PR DESCRIPTION
**Pull request by alainvd**
Prevention of duplicates

### CODE REVIEW

File: app/Services/PlatformUniqueIdService.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * @param int|null $platform_id
+     * @param mixed $puid
+     * @return bool
+     */
+    public function isPuidInCache(
+        ?int $platform_id,
+        mixed $puid
+    ): bool {
```
Comment (robertbrowncc) [inline app/Services/PlatformUniqueIdService.php:108]: Maybe we do just a return Cache:get(key, false);

File: database/migrations/2024_03_01_095414_create_archived_statements_table.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        Schema::create('archived_statements', function (Blueprint $table) {
+            $table->id();
+            $table->integer('platform_id');
+            $table->string('puid', 500);
+            $table->string('uuid', 36)->nullable();
+            $table->timestamp('date_received')->nullable();
+            $table->timestamps();
+
+            $table->unique(['platform_id', 'puid'], 'platform_puid_unique');
```
Comment (robertbrowncc) [inline database/migrations/2024_03_01_095414_create_archived_statements_table.php:22]: let's just go with index, if do unique we recreate the problem but in another table.

File: app/Services/PlatformUniqueIdService.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * @param $platform_id
+     * @param $puid
+     * @return void
+     * @throws PuidNotUniqueSingleException
+     */
+    public function addPuidToCache($platform_id, $puid): void
+    {
+        $key = $this->getCacheKey($platform_id, $puid);
+        if (Cache::has($key)) throw new PuidNotUniqueSingleException($puid);
```
Comment (robertbrowncc) [inline app/Services/PlatformUniqueIdService.php:44]: Maybe: if Cache:get(key, false) ?

File: app/Services/PlatformUniqueIdService.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        Log::info("add to cache: " . $key);
+        Cache::put($key, 0, now()->addDays($this->cache_valid_days));
```
Comment (robertbrowncc) [inline app/Services/PlatformUniqueIdService.php:45]: Put a true

File: app/Services/PlatformUniqueIdService.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            ArchivedStatement::create([
+                'puid' => $puid,
+                'date_received' => Carbon::now(),
+                'platform_id' => $platform_id
+            ]);
+        } catch (QueryException $e) {
+            if (
+                str_contains($e->getMessage(), "platform_puid_unique") || // mysql
+                str_contains($e->getMessage(),
+                    "UNIQUE constraint failed: archived_statements.platform_id, archived_statements.puid") // sqlite
```
Comment (robertbrowncc) [inline app/Services/PlatformUniqueIdService.php:66]: maybe just a select and use a normal index.

File: app/Services/PlatformUniqueIdService.php [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        if (!empty($duplicates)) {
+            throw new PuidNotUniqueMultipleException($duplicates);
+        }
+    }
+
+    public function refreshPuidsInCache(array $puids, $platform_id)
+    {
+        foreach($puids as $puid){
+            $key = $this->getCacheKey($platform_id, $puid);
+            Cache::put($key, 0, now()->addDays($this->cache_valid_days));
```
Comment (robertbrowncc) [inline app/Services/PlatformUniqueIdService.php:151]: true instead of 0

Comment (alainvd) [reply app/Services/PlatformUniqueIdService.php]: set to true now

File: app/Services/PlatformUniqueIdService.php (hunk 5 above)
Comment (alainvd) [reply app/Services/PlatformUniqueIdService.php]: good idea. the index is not unique anymore.

File: app/Services/PlatformUniqueIdService.php (hunk 4 above)
Comment (alainvd) [reply app/Services/PlatformUniqueIdService.php]: done

File: app/Services/PlatformUniqueIdService.php (hunk 3 above)
Comment (alainvd) [reply app/Services/PlatformUniqueIdService.php]: sure, good idea.

File: database/migrations/2024_03_01_095414_create_archived_statements_table.php (hunk 2 above)
Comment (alainvd) [reply database/migrations/2024_03_01_095414_create_archived_statements_table.php]: yes !

File: app/Services/PlatformUniqueIdService.php (hunk 1 above)
Comment (alainvd) [reply app/Services/PlatformUniqueIdService.php:99]: sure, good idea.

File: app/Exceptions/PuidNotUniqueSingleException.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    public function getJsonResponse(): JsonResponse
+    {
+        $errors = [
+            'puid' => [
+                $this->getMessage(),
+            ],
+        ];
+
+        $response = ['message' => $this->getMessage(), 'errors' => $errors, 'existing' => (object)['puid' => $this->duplicate]];
```
Comment (robertbrowncc) [inline app/Exceptions/PuidNotUniqueSingleException.php:30]: Do we need the "object" there?

**Review by robertbrowncc — APPROVED** [review]

**Pull request merged**